### PR TITLE
Bytes sort

### PR DESCRIPTION
**Pull request by Mocramis**
I noticed i couldn't  sort operations on bytes field.

I do not know if the tests are enough, but this is a quite simple patch.

### CODE REVIEW

File: mongomock/filtering.py [hunk 1]
```diff
@@ -401,11 +405,24 @@ def operator_eq(doc_val, search_val):
     return operator.eq(doc_val, search_val)
 
 
+def cmp_operator(op, a, b):
```
Comment (pcorpet) [inline mongomock/filtering.py]: Please treat this as a special case in the bson_compare function. Or did I miss something?

File: tests/test__collection_api.py [hunk 2]
```diff
@@ -3740,9 +3740,12 @@ def test__filter_objects_comparison(self):
             {'_id': 11, 'counts': {'circles': mongomock.ObjectId()}},
             # Document kept: datetimes are more than numbers.
             {'_id': 12, 'counts': {'circles': datetime.now()}},
+            # Document kept: BinData are more than numbers.
+            {'_id': 13, 'counts': {'circles': b'binary'}},
         ])
         results = collection.find(query)
-        self.assertEqual({1, 2, 5, 9, 11, 12}, {doc['_id'] for doc in results})
+        self.assertEqual({1, 2, 5, 9, 11, 12, 13},
```
Comment (pcorpet) [inline tests/test__collection_api.py]: Please break line after `(`

File: mongomock/filtering.py [hunk 3]
```diff
@@ -401,11 +405,24 @@ def operator_eq(doc_val, search_val):
...
+    # pymongo handles bytes as binary in python3+:
```
Comment (pcorpet) [inline mongomock/filtering.py]: It's actually bson, not pymongo

**Review by pcorpet — CHANGES_REQUESTED** [review]
Very cool! Thanks a lot for this. And thanks for all the pointer to the code or documentation.

File: mongomock/filtering.py (hunk 1 above)
Comment (Mocramis) [reply mongomock/filtering.py]: this only apply to sorting operators. appart from checkimg wether an operator is one of the sorting operator in bson_compare, i'm not sure how to do it, and the former method does not seem better to me than having the correct operators right offt the bat.

File: mongomock/filtering.py (hunk 3 above)
Comment (Mocramis) [reply mongomock/filtering.py]: my bad

File: mongomock/filtering.py (hunk 1 above)
Comment (pcorpet) [reply mongomock/filtering.py]: I'm not sure to understand what you're implying. `bson_compare` is only ever called with values from the `SORTING_OPERATOR_MAP` dict.

I think that comparing two bytes fields even when not sorting should use your method. `{'$lt': [b'abc', b'b']}`.

I suggest you keep the tests you have added and try moving the code to bson_compare: they should pass and not break any other tests.

Comment (Mocramis) [reply mongomock/filtering.py]: my bad dunno why, i though it was called for other operators. done.

File: mongomock/filtering.py [hunk 4]
```diff
@@ -1,4 +1,5 @@
 from datetime import datetime
+from functools import partial
```
Comment (pcorpet) [inline mongomock/filtering.py]: Unused.

File: mongomock/filtering.py [hunk 5]
```diff
@@ -287,6 +288,16 @@ def bson_compare(op, a, b, can_compare_types=True):
...
         return op(0, 0)
 
+    # bson handles bytes as binary in python3+:
+    # https://api.mongodb.com/python/current/api/bson/index.html
+    if PY3 and isinstance(a, bytes):
+        # Performs the same operation as described by:
+        # https://docs.mongodb.com/manual/reference/bson-type-comparison-order/#bindata
+        if len(a) != len(b):
+            return op(len(a), len(b))
+        # bytes is always treated as subtype 0 by the bson library
```
Comment (pcorpet) [inline mongomock/filtering.py:297]: Drop those two lines, as the default is to run it in this function (see line 301).

Comment (Mocramis) [reply mongomock/filtering.py:297]: nope if  i remove these, the ordering differs with the mongo one (with a slight change to the tests that I am going to push).

Comment (Mocramis) [reply mongomock/filtering.py:297]: with current tests , i get (without these two lines):

```
======================================================================
FAIL: test__find_sort_list_empty_order (tests.test__mongomock.MongoClientCollectionTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/data/mongomock/tests/test__mongomock.py", line 829, in test__find_sort_list_empty_order
    self.cmp.compare.find(sort=[('b', 1)])
  File "/data/mongomock/tests/multicollection.py", line 70, in __call__
    _assert_no_diff(results, ignore_order=self.___ignore_order, sort_by=self.___sort_by)
  File "/data/mongomock/tests/multicollection.py", line 95, in _assert_no_diff
    assert not d, _format_diff_message(prev_name, name, d)
AssertionError: Unexpected Diff:
	fake.3.b != real.3.b (b'alongbin1' != b'bin1')
	fake.4.b != real.4.b (b'alongbin2' != b'bin2')
	fake.5.b != real.5.b (b'bin1' != b'alongbin1')
	fake.6.b != real.6.b (b'bin2' != b'alongbin2')
```

Comment (pcorpet) [reply mongomock/filtering.py:297]: My point is that if you remove `'return op(a, b)`, Python will leave the `if PY3 and isinstance(a, bytes)` branch and go to the last line of the function which is also `return op(a, b)`, so I don't see how it goes wrong.

Comment (pcorpet) [reply mongomock/filtering.py:297]: Basically:

```py
if a:
  if b:
     return 0
  return 1
return 1
```

is equivalent to 

```py
if a:
  if b:
     return 0
return 1
```

Comment (Mocramis) [reply mongomock/filtering.py:297]: Ha ok. Dunno if it is due to an update, but the quoted code included the two lines before (and not the return) so i though you were talking about the 
```
if len(a) != len(b):
   return op(len(a), len(b))
```
lines

Sorry, should have guessed it. Will do !

**Review by pcorpet — APPROVED** [review]

**Pull request merged**